src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc

```diff
@@ -1570,23 +1635,89 @@ grpc_channel_args* GrpcLb::CreateRoundRobinPolicyArgsLocked() {
   return args;
 }
 
-void GrpcLb::CreateOrUpdateRoundRobinPolicyLocked() {
+void GrpcLb::CreateChildPolicyLocked(const char* name,
+                                     grpc_channel_args* args) {
+  OrphanablePtr<LoadBalancingPolicy>& lb_policy =
+      child_policy_ == nullptr ? child_policy_ : pending_child_policy_;
+  Helper* helper = New<Helper>(Ref());
+  LoadBalancingPolicy::Args lb_policy_args;
+  lb_policy_args.combiner = combiner();
+  lb_policy_args.args = args;
+  lb_policy_args.channel_control_helper =
+      UniquePtr<ChannelControlHelper>(helper);
+  lb_policy = LoadBalancingPolicyRegistry::CreateLoadBalancingPolicy(
+      name, child_policy_config_, std::move(lb_policy_args));
+  if (GPR_UNLIKELY(lb_policy == nullptr)) {
+    gpr_log(GPR_ERROR, "[grpclb %p] Failure creating child policy %s", this,
+            name);
+    return;
+  }
+  helper->set_child(lb_policy.get());
+  if (grpc_lb_glb_trace.enabled()) {
+    gpr_log(GPR_INFO, "[grpclb %p] Created new child policy %s (%p)", this,
+            name, lb_policy.get());
+  }
+  // Add the gRPC LB's interested_parties pollset_set to that of the newly
+  // created child policy. This will make the child policy progress upon
+  // activity on gRPC LB, which in turn is tied to the application's call.
+  grpc_pollset_set_add_pollset_set(lb_policy->interested_parties(),
+                                   interested_parties());
+  lb_policy->ExitIdleLocked();
+}
+
+void GrpcLb::CreateOrUpdateChildPolicyLocked() {
   if (shutting_down_) return;
-  grpc_channel_args* args = CreateRoundRobinPolicyArgsLocked();
+  grpc_channel_args* args = CreateChildPolicyArgsLocked();
   GPR_ASSERT(args != nullptr);
-  if (rr_policy_ != nullptr) {
+  // There are several cases here:
+  // 1. If we do not yet have any child policy, we create one and store
+  //    it in child_policy_.
+  // 2. If we have a child policy but no pending child policy, then:
+  //    a. If the child policy's name equals child_policy_name, then we
+  //       update the existing child policy.
+  //    b. If the child policy's name does not equal child_policy_name,
+  //       we create a new policy.  The policy will be stored in
+  //       pending_child_policy_ and will later be swapped into
+  //       child_policy_ by the helper when the new child transitions
+  //       into state READY.
+  // 3. If we have both a child policy and a pending child policy (i.e.,
+  //    we created a new policy for a previous update that has not yet
+  //    been moved from pending_child_policy_ to child_policy_), then:
+  //    a. If the pending child policy's name equals child_policy_name,
```

Comment:
What if `child_policy_name` equals the the current child policy's name?

Reply:
I thought about that, but that case should be sufficiently rare that I didn't feel the need to optimize it.  I think the only way that could happen is if we change from the service config from LB policy A to LB policy B and then very quickly (before the client can complete the switch from A to B) switch back to A.  In that situation, I think it's fine to treat it like a new update and create a new policy.

Keep in mind that in the common case, due to the subchannel pool, most child policies will switch almost instantaneously, because if there's a single READY subchannel under the old policy, the new policy will transition to state READY right away.  The only cases where that would not be true would be if there were no READY subchannels when the LB policy switches or if the list of subchannels changes to a completely disjoint set in the same update that gives us the new service config.  We could optimize this slightly by immediately swapping out the LB policy if the old policy is not currently in state READY, but again, it's not obvious that it's worth the complexity to do that right now.  We can always add it later if it turns out to be important.

Reply:
I propose this:
```
  // There are several cases here:
  // 1. If we do not yet have any LB policy, we create one and store
  //    it in lb_policy_.
  // 2. If we have an LB policy, whose name equals lb_policy_name, then we
  //    update the existing LB policy. If there is a pending LB policy, reset 
  //    it.
  // 3. If we have an LB policy, whose name does not equal lb_policy_name, then:
  //    a. If we do not yet have any pending LB policy, we create a new policy.
  //    The policy will be stored in pending_lb_policy_ and will later be
  //    swapped into lb_policy_ by the helper when the new LB transitions into
  //    state READY.
  //    b. If we have a pending LB policy, whose name equals lb_policy_name,
  //    then we update the existing pending LB policy.
  //    c. If we have a pending LB policy, whose name does not equal
  //    lb_policy_name, then we create a new policy. The new policy is stored in
  //    pending_lb_policy_ (replacing the one that was there before, which will
  //    be immediately shut down) and will later be swapped into lb_policy_ by
  //    the helper when the new LB transitions into state READY.
```

Reply:
It's true that A->B->A update is rare, but I think the code won't be a lot more complex with that optimization. Well, I GUESS so, given the new branch discussion above. I will try implementing it concisely.

Reply:
Even with this optimization, the code is not complex or long. So I think it's worth adding.

https://github.com/grpc/grpc/pull/18101/files#diff-19b9841b5d34408bb5328232a51f68d5R288

Reply:
I agree that the individual change you're suggesting is not large, but I think this code is already fairly complex, and I think adding this case makes it harder to understand.  I would prefer to leave this out.

To be honest, if I was going to add any optimization here, I'd be more likely to add the optimization of immediately swapping in the new policy if the old policy is not in state READY.  That one is more likely to get used and IMHO is a bit easier to explain.

But neither of these is really necessary right now, so I'd prefer to stick to the basics.

Reply:
I still think it's worth adding, because:

1. It's an optimization, which is good to have.
2. Adding it and reorganizing the condition branches makes the logic flow easier to understand. 
   1. The way we organize the condition branches in the comment now matches the initialization of `create_policy` better. It's kind of exact match now... 
   2. It reduces the number of sub-conditions from 4 to 3.
3. This optimization is easier than the one to swap in the new policy if the old one is not in state READY, because it doesn't need to check any connectivity state. It's just checking the existence and the name of the LB policies, which are already checked even without this optimization.

Reply:
> 1. It's an optimization, which is good to have.

I don't think that optimizations are inherently good; I think that they need to be approached using a cost/benefit analysis, just like any other code change.  In this case, the benefit is very small, and while the complexity of the optimization itself is not huge, it is IMHO large enough to outweigh the benefit we would get by having it.

> 2. Adding it and reorganizing the condition branches makes the logic flow easier to understand.
>    
>    1. The way we organize the condition branches in the comment now matches the initialization of `create_policy` better. It's kind of exact match no ew...

I don't actually think it's important for the branches in the comment to match the structure of the code.  I think comments should be structured in a way that makes it clear to a human reader what is actually happening, and code should be structured to be efficient and maintainable.  Those differing sets of goals often mean that the two will not match up.

From a human reader's perspective, I think the important thing to understand is that there are basically 3 cases: (1) we don't yet have any LB policy, (2) we have a child policy but no pending child policy, and (3) we have both a child policy and a pending child policy.  The fact that cases (2) and (3) each have two branches is another level of detail.  That's why I structured the comment the way I did.

If I had wanted the code to be structured the same way that the comment is, I would have written something like this:

```
if (child_policy_ == nullptr) {
  // ...case 1...
} else if (pending_child_policy_ == nullptr) {
  // ...case 2...
} else {
  // ...case 3...
}
```

But that would have resulted in duplicate code between the three branches, which is why I didn't do that.

>    2. It reduces the number of sub-conditions from 4 to 3.

I don't see how it actually does this.  There are actually 5 cases either way.

> 3. This optimization is easier than the one to swap in the new policy if the old one is not in state READY, because it doesn't need to check any connectivity state. It's just checking the existence and the name of the LB policies, which are already checked even without this optimization.

That's true.  But notice that I am not adding either optimization here. :)